### PR TITLE
Refactor access controller

### PR DESCRIPTION
**Pull request by haadcode**
This PR refactors the access controller in a way that the ACL handles all keys, signing and verification related functionality. The general idea is that Log/Entry don't know anything about the keys or keystore and only gets and `acl` object (the access controller) which encapsulates the functionality.

I did the previous commits in a bit of a rush, so there may be some weirdness here. Would love to hear your feedback @coyotespike, @shamb0t and @vvp.

I'm unsure if the ACL api is what we want to do in the end, but at least this would give us a "clean" separation of the acl and the log.

I'll try to check this again myself in the next couple of weeks to make sure everything makes sense, but for now, I hope this helps you to proceed with the refactoring / acl work that you're doing.

### CODE REVIEW

**Comment by haadcode** [comment]
@fazo96 would love to hear your feedback too on this. Your PR #123 has been open for a long time (sorry about that! :/) and this one is based on yours but with a very different code as to how to do the custom verification. I believe this would still give you that functionality but with the difference that instead of passing a custom verification function, you'd have to write a custom access controller (which I think would make sense generally if you're planning to check the ACL against a smart contract).

**Comment by fazo96** [comment]
hey @haadcode, thanks for involving me in this :)

I think this ACL refactor looks much better designed than my PR and will get me the same functionality.

Also thrilled by how easy it looks to pass a custom key to the ACL :+1: 

One thing I can't wrap my head around is what happens if an ACL is not passed? No more signing of the entries by default? I'm probably just missing something

**Comment by haadcode** [comment]
> One thing I can't wrap my head around is what happens if an ACL is not passed? No more signing of the entries by default? I'm probably just missing something

I left out that part and we should discuss what the default behaviour should be. You're correct that right now the entries don't get signed if an acl is not passed (which prolly doesn't have a lot of use cases as it breaks the integrity of the log). I suppose we could either throw an error or defer to a default access controller (like it used to be / is now). Any thoughts?

@fazo96 do you have any links to how you're using the custom verification in your code? Would love to understand your use case better so that we can take that into consideration designing all this.

**Comment by fazo96** [comment]
@haadcode sure, here is the custom OrbitDB Store we use in the current [Chlu](https://chlu.io) implementation:

https://github.com/ChluNetwork/chlu-ipfs-support/blob/master/src/modules/orbitdb/store.js

For our validation needs, there's three points at which we can validate content before accepting it:

1. before joining logs (using either my PR #127 or this PR and a custom ACL)
2. in the `updateIndex` function, so we don't put invalid data in the index
3. when reading from the index
4. after reading from the index

We haven't much validation going on related to orbit-db data, but the ReviewRecords we store are supposed to be independently validatable objects and they are, so the validation for now occures in the code that comes __after reading from the orbit-db index (point 4)__.

This is not ideal because the index can get polluted by any fake information, and also we allow review records to be updated and the current implementation makes it possible for anyone to create a bogus update and thus making the review record invalid for everyone.

The point of my PR #127 was to make __point 1__ possible and https://github.com/orbitdb/orbit-db-store/pull/21 already made __point 2__ possible. __point 3__ is easy because we have a custom index and __point 4__ does not involve orbit-db.

So we have an external, async validation function that can be called with the CID of the review record (passed in the payload of the orbit-db operation). This function will return true if the data is valid and should be kept in orbit-db.

In conclusion, our use case was making sure we have the means of keeping invalid data out of the index or even out of the oplog, depends on the user/application of our Chlu protocol (maybe it is preferable to keep invalid data in the oplog for further analysis and debugging, for example, but still keep it out of the index).

Another important point in favour is that to authenticate users we use a [Decentralized Identifier (DID)](https://w3c-ccg.github.io/did-spec/) that has its own keypair and the custom ACL in this PR makes it easier for us to reuse DID keypairs for orbit-db signing and validating.

**Comment by coyotespike** [comment]
`ipfs-log` just takes an `acl` - so it would seem that we can still pass an acl down from `orbit-db` through `orbit-db-store`.

Do you intend `acl.js` to move to `orbit-db`? I am curious about the split of responsibilities.  The access controllers on the dynamic permissions branch are setting and getting capabilities. `acl.js` uses those keys, and the keystore, to check permissions with `sign` and `verify` etc.

The smart contract access controller needs to do both, because it can't provide a list of keys. I suppose we'll just pass down the smart contract access controller.

File: src/log.js [hunk 1]
```diff
@@ -198,20 +200,13 @@ class Log extends GSet {
    * @return {Log}   New Log containing the appended value
    */
   async append (data, pointerCount = 1) {
```
[the remark this answers is not in the document]
Comment (haadcode) [reply src/log.js]: Your previous comment was a good point though. I was thinking if we should make it (calling `canAppend` or `sign`) explicit and call it in the log instead of in Entry?

File: src/log.js [hunk 2]
```diff
@@ -475,15 +408,16 @@ class Log extends GSet {
    * @param {Number}              [length=-1]   How many entries to include. Default: infinite.
    * @param {Array<Entry|string>} [exclude]     Array of entries or hashes or entries to not fetch (foe eg. cached entries)
    * @param {Function(hash, entry, parent, depth)} [onProgressCallback]
+   * @param {Function} [verifyEntry]   Custom entry verification function
    * @return {Promise<Log>}       New Log
    */
-  static fromEntry (ipfs, sourceEntries, length = -1, exclude, onProgressCallback) {
+  static fromEntry (ipfs, sourceEntries, length = -1, exclude, onProgressCallback, verifyEntry) {
```
[the remark this answers is not in the document]
Comment (haadcode) [reply src/log.js]: Thank you! 👍Will fix all the `verifyEntry` args as per your comments.

File: src/log.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (coyotespike) [reply src/log.js]: I already forgot it a second time :joy: so we could either make it explicit, or make a generalized testing suite for the access controller API to ensure that all access controllers conform to this API.

File: src/entry.js [hunk 3]
```diff
@@ -43,23 +43,27 @@ class Entry {
...
+      entry = await Entry.sign(entry, acl) 
     }
 
     entry.hash = await Entry.toMultihash(ipfs, entry)
     return entry
   }
 
-  static async signEntry (keystore, entry, key) {
-    const signature = await keystore.sign(key, Buffer.from(JSON.stringify(entry)))
+  static async sign (entry, acl) {
```
Comment (coyotespike) [inline src/entry.js]: A smart contract access controller will want to augment the entry with a `chainSignature` and `chainPublicKey`, or something like that.

File: src/acl.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
+class ACL {
+  constructor (keystore, key, keys = []) {
+    this._keystore = keystore
```
Comment (thiagodelgado111) [inline src/acl.js]: I think here it should only get `sign`, `getPubKey` and `verify` functions instead of direct access to the keys

File: src/acl.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
+class ACL {
```
Comment (thiagodelgado111) [inline src/acl.js]: Maybe we could call it `IntegrityChecker` to avoid confusion around orbit-db and ipfs-log ACL

File: src/acl.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+  getSigningKey () {
+    return this._key
+  }
+
+  getPublicSigningKey (format = 'hex') {
+    return this._key.getPublic(format)
+  }
+
+  canAppend (signingKey) {
```
Comment (thiagodelgado111) [inline src/acl.js]: If we pass functions from orbit-db access controller to it, we wouldn't need this anymore

File: src/log.js [hunk 7]
```diff
@@ -198,20 +200,13 @@ class Log extends GSet {
...
-    }
-
     // Update the clock (find the latest clock)
     const newTime = Math.max(this.clock.time, this.heads.reduce(maxClockTimeReducer, 0)) + 1
     this._clock = new Clock(this.clock.id, newTime)
     // Get the required amount of hashes to next entries (as per current state of the log)
     const nexts = Object.keys(this.traverse(this.heads, pointerCount))
     // Create the entry and add it to the internal cache
-    const entry = await Entry.create(this._storage, this._keystore, this.id, data, nexts, this.clock, this._key)
+    const entry = await Entry.create(this._storage, this._acl, this.id, data, nexts, this.clock)
```
Comment (thiagodelgado111) [inline src/log.js]: Instead of passing the whole ACL object here, the `Entry.create` method could be broken up into create/sign/put. It might make it easier to understand how it actually works – since one might assume this is simply creating an entry instance – and then we can pass just the public key to be used as a fallback for the `clockId` property. Maybe a TODO for another PR?

File: src/acl.js [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    this._keystore = keystore
+    this._key = key
+    this._keys = keys && Array.isArray(keys) ? keys : keys ? [keys] : []
+  }
+
+  getSigningKey () {
+    return this._key
+  }
+
+  getPublicSigningKey (format = 'hex') {
```
Comment (thiagodelgado111) [inline src/acl.js]: I think `ipfs-log` always uses this format. Is it going to change or can we assume it'll always be 'hex'?

File: src/log.js [hunk 9]
```diff
@@ -578,6 +508,30 @@ class Log extends GSet {
     entries.forEach(addToIndex)
     return entries.reduce(reduceTailHashes, [])
   }
+
+  static difference (a, b) {
```
Comment (thiagodelgado111) [inline src/log.js]: Is it similar to `utils/difference`?

File: src/log.js [hunk 10]
```diff
@@ -57,10 +61,8 @@ class Log extends GSet {
     this._storage = ipfs
     this._id = id || randomId()
 
-    // Signing related setup
-    this._keystore = this._storage.keystore
-    this._key = key 
-    this._keys = Array.isArray(keys) ? keys : [keys]
+    // Access Controller
+    this._acl = acl
```
Comment (thiagodelgado111) [inline src/log.js]: Like @fazo96 mentioned in his comment (https://github.com/orbitdb/ipfs-log/pull/128#issuecomment-399921075), maybe we should add some validation here that checks if the ACL exists and if it follows the "interface" we need. If it doesn't, we throw an error. Is there a scenario where it wouldn't be necessary?

File: src/entry.js [hunk 11]
```diff
@@ -43,23 +43,27 @@ class Entry {
       clock: new Clock(clockId, clockTime),
     }
 
-    // If signing key was passedd, sign the enrty
-    if (keystore && signKey) {
-      entry = await Entry.signEntry(keystore, entry, signKey) 
+    // If acl was passed, sign the entry (=authorize)
+    if (acl) {
```
Comment (thiagodelgado111) [inline src/entry.js]: I think this should be a breaking change, providing an ACL should be mandatory from now on

**Comment by haadcode** [comment]
Closing as obsolete in favor of https://github.com/orbitdb/ipfs-log/pull/159.
